test/replication-luatest/gh_7086_box_issue_promote_assert_test.lua

```diff
@@ -0,0 +1,137 @@
+                server_1:wait_election_term(server_2:election_term())
+                server_1:wait_vclock_of(server_2)
+            end
+        end
+    end
+end
+
+local function promote(server)
+    luatest.helpers.retrying({}, function()
+        return server:exec(function()
+            pcall(box.ctl.promote)
+            if box.info.election.state ~= 'leader' then
+                return false
+            end
+            if box.info.synchro.queue.owner ~= box.info.id then
+                return false
+            end
+            return true
+        end)
+    end)
+    server:exec(function() box.ctl.wait_rw() end)
+end
+
+local function promote_start(server)
+    return server:exec(function()
+        local f = require('fiber').new(box.ctl.promote)
+        f:set_joinable(true)
+        return f:id()
+    end)
+end
+
+local function demote_start(server)
+    return server:exec(function()
+        local f = require('fiber').new(box.ctl.demote)
+        f:set_joinable(true)
+        return f:id()
+    end)
+end
+
+local function fiber_join(server, fiber)
+    return server:exec(function(fiber)
+        return require('fiber').find(fiber):join()
+    end, {fiber})
+end
+
+g.before_all(function(g)
+    g.cluster = cluster:new({})
+
+    g.box_cfg = {
+        election_mode = 'off',
+        replication_timeout = 0.1,
+        replication_synchro_timeout = 5,
+        replication_synchro_quorum = 1,
+        replication = {
+            server.build_instance_uri('server_1'),
+            server.build_instance_uri('server_2'),
+        },
+    }
+
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', box_cfg = g.box_cfg})
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', box_cfg = g.box_cfg})
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+end)
+
+-- Demoting should fail if it is interrupted from another server
+-- while acquiring limbo.
+g.test_txn_limbo_begin_interfering_promote = function(g)
+    g.server_2:exec(function()
+        box.error.injection.set('ERRINJ_TXN_LIMBO_BEGIN_DELAY', true)
+    end)
+
+    local term = g.server_1:election_term()
+    local f = promote_start(g.server_2)
+    g.server_1:wait_election_term(term + 1)
+
+    local term = g.server_2:election_term()
+    g.server_1:exec(function() box.ctl.promote() end)
+    g.server_2:wait_election_term(term + 1)
+
+    g.server_2:exec(function()
+        box.error.injection.set('ERRINJ_TXN_LIMBO_BEGIN_DELAY', false)
+    end)
+
+    local ok, err = fiber_join(g.server_2, f)
+    luatest.assert(not ok and
+        (err.code == box.error.INTERFERING_PROMOTE or
+            err.code == box.error.INTERFERING_ELECTIONS),
+        'interfering promote not handled')
+
+    g.server_1:exec(function() box.ctl.demote() end)
+    wait_sync(g.cluster.servers)
+end
+
+-- Demoting should fail if it is interrupted from another server
+-- while acquiring limbo.
+g.test_txn_limbo_begin_interfering_demote = function(g)
```

Comment:
Both test cases don't crash when I run them locally. They fail, but are supposed to crash. At least in debug build. Why do they just fail with "not ok"?

Reply:
I guess, you tried running this test without the added error injection. that way, the test will result in two possible outcomes:
1) by chance - all timings are correct and it will trigger those assertions and crash
2) it will promote/demote just fine, because term bumps/promotes arrived before/after, resulting in failed test, because i check for specific error.